### PR TITLE
Custom serialization for dynamically created namedtuple

### PR DESCRIPTION
**Pull request by dvadym**

### CODE REVIEW

File: pipeline_dp/combiners.py [hunk 1]
```diff
@@ -217,6 +217,28 @@ def metrics_names(self) -> List[str]:
...
+                               field_names: tuple) -> 'MetricsTuple':
+    """Creates namedtuple type with a custom serializer."""
+
+    # The custom serializer is required for supporting searialization of
+    # namedtuples in Apache Beam.
+    cache_key = (type_name, field_names)
+    named_tuple = _named_tuple_cache.get(cache_key)
+    if named_tuple is None:
+        named_tuple = collections.namedtuple(type_name, field_names)
+        named_tuple.__reduce__ = lambda self: (_get_or_create_named_tuple,
```
Comment (dvadym) [inline pipeline_dp/combiners.py:235]: Setting `__reduce__` defines custom serializer. 
Custom serialzier should return a tuple (`func_to_create_object_type`, arguments for this function)

File: pipeline_dp/combiners.py [hunk 2]
```diff
@@ -217,6 +217,28 @@ def metrics_names(self) -> List[str]:
...
+# Cache for namedtuple types. It is should be used only in
+# '_get_or_create_named_tuple()' function.
+_named_tuple_cache = {}
+
+
+def _get_or_create_named_tuple(type_name: str,
+                               field_names: tuple) -> 'MetricsTuple':
+    """Creates namedtuple type with a custom serializer."""
+
+    # The custom serializer is required for supporting searialization of
```
Comment (miracvbasaran) [inline pipeline_dp/combiners.py:229]: nit: "serialization"

File: tests/dp_engine_test.py [hunk 3]
```diff
@@ -755,11 +755,9 @@ def test_run_e2e_local(self):
 
         self.assertEqual(5, len(list(output)))
 
-    @unittest.skipIf(
-        sys.platform == "win32" or
-        (sys.version_info.minor <= 7 and sys.version_info.major == 3),
-        "There are some problems with PySpark setup on older python and Windows"
-    )
+    @unittest.skip("There are some problems with serialization in this test. "
+                   "Tests in private_spark_test.py work normaly so probablyt it"
```
Comment (miracvbasaran) [inline tests/dp_engine_test.py:759]: nit: "probably"

File: tests/dp_engine_test.py [hunk 4]
```diff
@@ -755,11 +755,9 @@ def test_run_e2e_local(self):
 
         self.assertEqual(5, len(list(output)))
 
-    @unittest.skipIf(
-        sys.platform == "win32" or
-        (sys.version_info.minor <= 7 and sys.version_info.major == 3),
-        "There are some problems with PySpark setup on older python and Windows"
-    )
+    @unittest.skip("There are some problems with serialization in this test. "
```
Comment (miracvbasaran) [inline tests/dp_engine_test.py:758]: I wonder if this is related to me not being able to extract the metrics (e.g., count, sum) in private_spark_test.py by treating the output of `dp_engine.aggregate()` as a named tuple. I was only able to reference them as a regular tuple (e.g., by calling `v[0]`).

I imagine we do want to use namedtuples in spark as well, so maybe something to look into?

**Review by miracvbasaran — APPROVED** [review]

File: tests/dp_engine_test.py (hunk 4 above)
Comment (dvadym) [reply tests/dp_engine_test.py:758]: Yeah, it might be related. Let's do in the next PRs

**Pull request merged**